feat: generate completion subcommand

plustik opened the pull request:
Fixes #1167 #1022 

As seen in [issue 1167](https://github.com/dandavison/delta/issues/1167) and [issue](https://github.com/dandavison/delta/issues/1022) the existing completion script for ZSH does not work right now.

One solution discussed [here](https://github.com/dandavison/delta/issues/1022#issuecomment-1232368478) is adding a subcommand to generate a completion file for ZSH/for all shells. I implemented this subcommand in these changes.
The completion file generated by the new subcommand fixes both issues 1167 and 1022.

This solution enables us to generate up-to-date completion files for multiple shells including ZSH, Bash and Fish when needed. I consequently removed the existing completion files from etc/completion. If you prefer keeping these files I could also just replace them with the scripts generated with the new commands.

maverick1872 commented:
I just updated #1167 with a basic completion that I finally got around to putting together that would solve the issue with it. That being said if they could be auto-generated that would be super nice for long term maintenance.

Might be valuable to include in the PR an example of what the generated completion files would look like if @dandavison were to accept it for ease of reviewing.

dandavison commented:
This looks great @plustik! I'll try it out/review  soon.

dandavison commented:
## delta --generate-completion bash
```
_delta() {
    local i cur prev opts cmd
    COMPREPLY=()
    cur="${COMP_WORDS[COMP_CWORD]}"
    prev="${COMP_WORDS[COMP_CWORD-1]}"
    cmd=""
    opts=""

    for i in ${COMP_WORDS[@]}
    do
        case "${cmd},${i}" in
            ",$1")
                cmd="delta"
                ;;
            *)
                ;;
        esac
    done

    case "${cmd}" in
        delta)
            opts="-n -s -w -h -V --blame-code-style --blame-format --blame-palette --blame-separator-format --blame-separator-style --blame-timestamp-format --blame-timestamp-output-format --color-only --config --commit-decoration-style --commit-regex --commit-style --dark --default-language --diff-highlight --diff-so-fancy --diff-stat-align-width --features --file-added-label --file-copied-label --file-decoration-style --file-modified-label --file-removed-label --file-renamed-label --file-style --file-transformation --generate-completion --grep-context-line-style --grep-file-style --grep-header-decoration-style --grep-header-file-style --grep-line-number-style --grep-output-type --grep-match-line-style --grep-match-word-style --grep-separator-symbol --hunk-header-decoration-style --hunk-header-file-style --hunk-header-line-number-style --hunk-header-style --hunk-label --hyperlinks --hyperlinks-commit-link-format --hyperlinks-file-link-format --inline-hint-style --inspect-raw-lines --keep-plus-minus-markers --light --line-buffer-size --line-fill-method --line-numbers --line-numbers-left-format --line-numbers-left-style --line-numbers-minus-style --line-numbers-plus-style --line-numbers-right-format --line-numbers-right-style --line-numbers-zero-style --list-languages --list-syntax-themes --map-styles --max-line-distance --max-line-length --merge-conflict-begin-symbol --merge-conflict-end-symbol --merge-conflict-ours-diff-header-decoration-style --merge-conflict-ours-diff-header-style --merge-conflict-theirs-diff-header-decoration-style --merge-conflict-theirs-diff-header-style --minus-empty-line-marker-style --minus-emph-style --minus-non-emph-style --minus-style --navigate --navigate-regex --no-gitconfig --pager --paging --parse-ansi --plus-emph-style --plus-empty-line-marker-style --plus-non-emph-style --plus-style --raw --relative-paths --right-arrow --show-colors --show-config --show-syntax-themes --show-themes --side-by-side --syntax-theme --tabs --true-color --whitespace-error-style --width --word-diff-regex --wrap-left-symbol --wrap-max-lines --wrap-right-percent --wrap-right-prefix-symbol --wrap-right-symbol --zero-style --24-bit-color --help --version [MINUS_FILE] [PLUS_FILE]"
            if [[ ${cur} == -* || ${COMP_CWORD} -eq 1 ]] ; then
                COMPREPLY=( $(compgen -W "${opts}" -- "${cur}") )
                return 0
            fi
            case "${prev}" in
                --blame-code-style)
                    COMPREPLY=($(compgen -f "${cur}"))
                    return 0
                    ;;
                --blame-format)
                    COMPREPLY=($(compgen -f "${cur}"))
                    return 0
                    ;;
                --blame-palette)
                    COMPREPLY=($(compgen -f "${cur}"))
                    return 0
                    ;;
                --blame-separator-format)
                    COMPREPLY=($(compgen -f "${cur}"))
                    return 0
                    ;;
                --blame-separator-style)
                    COMPREPLY=($(compgen -f "${cur}"))
                    return 0
                    ;;
                --blame-timestamp-format)
                    COMPREPLY=($(compgen -f "${cur}"))
                    return 0
                    ;;
                --blame-timestamp-output-format)
                    COMPREPLY=($(compgen -f "${cur}"))
                    return 0
                    ;;
                --config)
                    COMPREPLY=($(compgen -f "${cur}"))
                    return 0
                    ;;
                --commit-decoration-style)
                    COMPREPLY=($(compgen -f "${cur}"))
                    return 0
                    ;;
                --commit-regex)
                    COMPREPLY=($(compgen -f "${cur}"))
                    return 0
                    ;;
                --commit-style)
                    COMPREPLY=($(compgen -f "${cur}"))
                    return 0
                    ;;
                --default-language)
                    COMPREPLY=($(compgen -f "${cur}"))
                    return 0
                    ;;
                --diff-stat-align-width)
                    COMPREPLY=($(compgen -f "${cur}"))
                    return 0
                    ;;
                --features)
                    COMPREPLY=($(compgen -f "${cur}"))
                    return 0
                    ;;
                --file-added-label)
                    COMPREPLY=($(compgen -f "${cur}"))
                    return 0
                    ;;
                --file-copied-label)
                    COMPREPLY=($(compgen -f "${cur}"))
                    return 0
                    ;;
                --file-decoration-style)
                    COMPREPLY=($(compgen -f "${cur}"))
                    return 0
                    ;;
                --file-modified-label)
                    COMPREPLY=($(compgen -f "${cur}"))
                    return 0
                    ;;
                --file-removed-label)
                    COMPREPLY=($(compgen -f "${cur}"))
                    return 0
                    ;;
                --file-renamed-label)
                    COMPREPLY=($(compgen -f "${cur}"))
                    return 0
                    ;;
                --file-style)
                    COMPREPLY=($(compgen -f "${cur}"))
                    return 0
                    ;;
                --file-transformation)
                    COMPREPLY=($(compgen -f "${cur}"))
                    return 0
                    ;;
                --generate-completion)
                    COMPREPLY=($(compgen -W "bash elvish fish powershell zsh" -- "${cur}"))
                    return 0
                    ;;
                --grep-context-line-style)
                    COMPREPLY=($(compgen -f "${cur}"))
                    return 0
                    ;;
                --grep-file-style)
                    COMPREPLY=($(compgen -f "${cur}"))
                    return 0
                    ;;
                --grep-header-decoration-style)
                    COMPREPLY=($(compgen -f "${cur}"))
                    return 0
                    ;;
                --grep-header-file-style)
                    COMPREPLY=($(compgen -f "${cur}"))
                    return 0
                    ;;
                --grep-line-number-style)
                    COMPREPLY=($(compgen -f "${cur}"))
                    return 0
                    ;;
                --grep-output-type)
                    COMPREPLY=($(compgen -f "${cur}"))
                    return 0
                    ;;
                --grep-match-line-style)
                    COMPREPLY=($(compgen -f "${cur}"))
                    return 0
                    ;;
                --grep-match-word-style)
                    COMPREPLY=($(compgen -f "${cur}"))
                    return 0
                    ;;
                --grep-separator-symbol)
                    COMPREPLY=($(compgen -f "${cur}"))
                    return 0
                    ;;
                --hunk-header-decoration-style)
                    COMPREPLY=($(compgen -f "${cur}"))
                    return 0
                    ;;
                --hunk-header-file-style)
                    COMPREPLY=($(compgen -f "${cur}"))
                    return 0
                    ;;
                --hunk-header-line-number-style)
                    COMPREPLY=($(compgen -f "${cur}"))
                    return 0
                    ;;
                --hunk-header-style)
                    COMPREPLY=($(compgen -f "${cur}"))
                    return 0
                    ;;
                --hunk-label)
                    COMPREPLY=($(compgen -f "${cur}"))
                    return 0
                    ;;
                --hyperlinks-commit-link-format)
                    COMPREPLY=($(compgen -f "${cur}"))
                    return 0
                    ;;
                --hyperlinks-file-link-format)
                    COMPREPLY=($(compgen -f "${cur}"))
                    return 0
                    ;;
                --inline-hint-style)
                    COMPREPLY=($(compgen -f "${cur}"))
                    return 0
                    ;;
                --inspect-raw-lines)
                    COMPREPLY=($(compgen -f "${cur}"))
                    return 0
                    ;;
                --line-buffer-size)
                    COMPREPLY=($(compgen -f "${cur}"))
                    return 0
                    ;;
                --line-fill-method)
                    COMPREPLY=($(compgen -f "${cur}"))
                    return 0
                    ;;
                --line-numbers-left-format)
                    COMPREPLY=($(compgen -f "${cur}"))
                    return 0
                    ;;
                --line-numbers-left-style)
                    COMPREPLY=($(compgen -f "${cur}"))
                    return 0
                    ;;
                --line-numbers-minus-style)
                    COMPREPLY=($(compgen -f "${cur}"))
                    return 0
                    ;;
                --line-numbers-plus-style)
                    COMPREPLY=($(compgen -f "${cur}"))
                    return 0
                    ;;
                --line-numbers-right-format)
                    COMPREPLY=($(compgen -f "${cur}"))
                    return 0
                    ;;
                --line-numbers-right-style)
                    COMPREPLY=($(compgen -f "${cur}"))
                    return 0
                    ;;
                --line-numbers-zero-style)
                    COMPREPLY=($(compgen -f "${cur}"))
                    return 0
                    ;;
                --map-styles)
                    COMPREPLY=($(compgen -f "${cur}"))
                    return 0
                    ;;
                --max-line-distance)
                    COMPREPLY=($(compgen -f "${cur}"))
                    return 0
                    ;;
                --max-line-length)
                    COMPREPLY=($(compgen -f "${cur}"))
                    return 0
                    ;;
                --merge-conflict-begin-symbol)
                    COMPREPLY=($(compgen -f "${cur}"))
                    return 0
                    ;;
                --merge-conflict-end-symbol)
                    COMPREPLY=($(compgen -f "${cur}"))
                    return 0
                    ;;
                --merge-conflict-ours-diff-header-decoration-style)
                    COMPREPLY=($(compgen -f "${cur}"))
                    return 0
                    ;;
                --merge-conflict-ours-diff-header-style)
                    COMPREPLY=($(compgen -f "${cur}"))
                    return 0
                    ;;
                --merge-conflict-theirs-diff-header-decoration-style)
                    COMPREPLY=($(compgen -f "${cur}"))
                    return 0
                    ;;
                --merge-conflict-theirs-diff-header-style)
                    COMPREPLY=($(compgen -f "${cur}"))
                    return 0
                    ;;
                --minus-empty-line-marker-style)
                    COMPREPLY=($(compgen -f "${cur}"))
                    return 0
                    ;;
                --minus-emph-style)
                    COMPREPLY=($(compgen -f "${cur}"))
                    return 0
                    ;;
                --minus-non-emph-style)
                    COMPREPLY=($(compgen -f "${cur}"))
                    return 0
                    ;;
                --minus-style)
                    COMPREPLY=($(compgen -f "${cur}"))
                    return 0
                    ;;
                --navigate-regex)
                    COMPREPLY=($(compgen -f "${cur}"))
                    return 0
                    ;;
                --pager)
                    COMPREPLY=($(compgen -f "${cur}"))
                    return 0
                    ;;
                --paging)
                    COMPREPLY=($(compgen -f "${cur}"))
                    return 0
                    ;;
                --plus-emph-style)
                    COMPREPLY=($(compgen -f "${cur}"))
                    return 0
                    ;;
                --plus-empty-line-marker-style)
                    COMPREPLY=($(compgen -f "${cur}"))
                    return 0
                    ;;
                --plus-non-emph-style)
                    COMPREPLY=($(compgen -f "${cur}"))
                    return 0
                    ;;
                --plus-style)
                    COMPREPLY=($(compgen -f "${cur}"))
                    return 0
                    ;;
                --right-arrow)
                    COMPREPLY=($(compgen -f "${cur}"))
                    return 0
                    ;;
                --syntax-theme)
                    COMPREPLY=($(compgen -f "${cur}"))
                    return 0
                    ;;
                --tabs)
                    COMPREPLY=($(compgen -f "${cur}"))
                    return 0
                    ;;
                --true-color)
                    COMPREPLY=($(compgen -f "${cur}"))
                    return 0
                    ;;
                --whitespace-error-style)
                    COMPREPLY=($(compgen -f "${cur}"))
                    return 0
                    ;;
                --width)
                    COMPREPLY=($(compgen -f "${cur}"))
                    return 0
                    ;;
                -w)
                    COMPREPLY=($(compgen -f "${cur}"))
                    return 0
                    ;;
                --word-diff-regex)
                    COMPREPLY=($(compgen -f "${cur}"))
                    return 0
                    ;;
                --wrap-left-symbol)
                    COMPREPLY=($(compgen -f "${cur}"))
                    return 0
                    ;;
                --wrap-max-lines)
                    COMPREPLY=($(compgen -f "${cur}"))
                    return 0
                    ;;
                --wrap-right-percent)
                    COMPREPLY=($(compgen -f "${cur}"))
                    return 0
                    ;;
                --wrap-right-prefix-symbol)
                    COMPREPLY=($(compgen -f "${cur}"))
                    return 0
                    ;;
                --wrap-right-symbol)
                    COMPREPLY=($(compgen -f "${cur}"))
                    return 0
                    ;;
                --zero-style)
                    COMPREPLY=($(compgen -f "${cur}"))
                    return 0
                    ;;
                --24-bit-color)
                    COMPREPLY=($(compgen -f "${cur}"))
                    return 0
                    ;;
                *)
                    COMPREPLY=()
                    ;;
            esac
            COMPREPLY=( $(compgen -W "${opts}" -- "${cur}") )
            return 0
            ;;
    esac
}

complete -F _delta -o nosort -o bashdefault -o default delta
```

## delta --generate-completion zsh
```
#compdef delta

autoload -U is-at-least

_delta() {
    typeset -A opt_args
    typeset -a _arguments_options
    local ret=1

    if is-at-least 5.2; then
        _arguments_options=(-s -S -C)
    else
        _arguments_options=(-s -C)
    fi

    local context curcontext="$curcontext" state line
    _arguments "${_arguments_options[@]}" \
'--blame-code-style=[Style string for the code section of a git blame line]:STYLE: ' \
'--blame-format=[Format string for git blame commit metadata]:FMT: ' \
'--blame-palette=[Background colors used for git blame lines (space-separated string)]:COLORS: ' \
'--blame-separator-format=[Separator between the blame format and the code section of a git blame line]:FMT: ' \
'--blame-separator-style=[Style string for the blame-separator-format]:STYLE: ' \
'--blame-timestamp-format=[Format of \`git blame\` timestamp in raw git output received by delta]:FMT: ' \
'--blame-timestamp-output-format=[Format string for git blame timestamp output]:FMT: ' \
'--config=[Load the config file at PATH instead of ~/.gitconfig]:PATH: ' \
'--commit-decoration-style=[Style string for the commit hash decoration]:STYLE: ' \
'--commit-regex=[Regular expression used to identify the commit line when parsing git output]:REGEX: ' \
'--commit-style=[Style string for the commit hash line]:STYLE: ' \
'--default-language=[Default language used for syntax highlighting]:LANG: ' \
'--diff-stat-align-width=[Width allocated for file paths in a diff stat section]:N: ' \
'--features=[Names of delta features to activate (space-separated)]:FEATURES: ' \
'--file-added-label=[Text to display before an added file path]:STRING: ' \
'--file-copied-label=[Text to display before a copied file path]:STRING: ' \
'--file-decoration-style=[Style string for the file decoration]:STYLE: ' \
'--file-modified-label=[Text to display before a modified file path]:STRING: ' \
'--file-removed-label=[Text to display before a removed file path]:STRING: ' \
'--file-renamed-label=[Text to display before a renamed file path]:STRING: ' \
'--file-style=[Style string for the file section]:STYLE: ' \
'--file-transformation=[Sed-style command transforming file paths for display]:SED_CMD: ' \
'--generate-completion=[Print completion file for the given shell]:GENERATE_COMPLETION:(bash elvish fish powershell zsh)' \
'--grep-context-line-style=[Style string for non-matching lines of grep output]:STYLE: ' \
'--grep-file-style=[Style string for file paths in grep output]:STYLE: ' \
'--grep-header-decoration-style=[Style string for the header decoration in grep output]:STYLE: ' \
'--grep-header-file-style=[Style string for the file path part of the header in grep output]:STYLE: ' \
'--grep-line-number-style=[Style string for line numbers in grep output]:STYLE: ' \
'--grep-output-type=[Grep output format. Possible values\: "ripgrep" - file name printed once, followed by matching lines within that file, each preceded by a line number. "classic" - file name\:line number, followed by matching line. Default is "ripgrep" if \`rg --json\` format is detected, otherwise "classic"]:OUTPUT_TYPE: ' \
'--grep-match-line-style=[Style string for matching lines of grep output]:STYLE: ' \
'--grep-match-word-style=[Style string for the matching substrings within a matching line of grep output]:STYLE: ' \
'--grep-separator-symbol=[Separator symbol printed after the file path and line number in grep output]:STRING: ' \
'--hunk-header-decoration-style=[Style string for the hunk-header decoration]:STYLE: ' \
'--hunk-header-file-style=[Style string for the file path part of the hunk-header]:STYLE: ' \
'--hunk-header-line-number-style=[Style string for the line number part of the hunk-header]:STYLE: ' \
'--hunk-header-style=[Style string for the hunk-header]:STYLE: ' \
'--hunk-label=[Text to display before a hunk header]:STRING: ' \
'--hyperlinks-commit-link-format=[Format string for commit hyperlinks (requires --hyperlinks)]:FMT: ' \
'--hyperlinks-file-link-format=[Format string for file hyperlinks (requires --hyperlinks)]:FMT: ' \
'--inline-hint-style=[Style string for short inline hint text]:STYLE: ' \
'--inspect-raw-lines=[Kill-switch for --color-moved support]:true|false: ' \
'--line-buffer-size=[Size of internal line buffer]:N: ' \
'--line-fill-method=[Line-fill method in side-by-side mode]:STRING: ' \
'--line-numbers-left-format=[Format string for the left column of line numbers]:FMT: ' \
'--line-numbers-left-style=[Style string for the left column of line numbers]:STYLE: ' \
'--line-numbers-minus-style=[Style string for line numbers in the old (minus) version of the file]:STYLE: ' \
'--line-numbers-plus-style=[Style string for line numbers in the new (plus) version of the file]:STYLE: ' \
'--line-numbers-right-format=[Format string for the right column of line numbers]:FMT: ' \
'--line-numbers-right-style=[Style string for the right column of line numbers]:STYLE: ' \
'--line-numbers-zero-style=[Style string for line numbers in unchanged (zero) lines]:STYLE: ' \
'--map-styles=[Map styles encountered in raw input to desired output styles]:STYLES_MAP: ' \
'--max-line-distance=[Maximum line pair distance parameter in within-line diff algorithm]:DIST: ' \
'--max-line-length=[Truncate lines longer than this]:N: ' \
'--merge-conflict-begin-symbol=[String marking the beginning of a merge conflict region]:STRING: ' \
'--merge-conflict-end-symbol=[String marking the end of a merge conflict region]:STRING: ' \
'--merge-conflict-ours-diff-header-decoration-style=[Style string for the decoration of the header above the '\''ours'\'' merge conflict diff]:STYLE: ' \
'--merge-conflict-ours-diff-header-style=[Style string for the header above the '\''ours'\'' branch merge conflict diff]:STYLE: ' \
'--merge-conflict-theirs-diff-header-decoration-style=[Style string for the decoration of the header above the '\''theirs'\'' merge conflict diff]:STYLE: ' \
'--merge-conflict-theirs-diff-header-style=[Style string for the header above the '\''theirs'\'' branch merge conflict diff]:STYLE: ' \
'--minus-empty-line-marker-style=[Style string for removed empty line marker]:STYLE: ' \
'--minus-emph-style=[Style string for emphasized sections of removed lines]:STYLE: ' \
'--minus-non-emph-style=[Style string for non-emphasized sections of removed lines that have an emphasized section]:STYLE: ' \
'--minus-style=[Style string for removed lines]:STYLE: ' \
'--navigate-regex=[Regular expression defining navigation stop points]:REGEX: ' \
'--pager=[Which pager to use]:CMD: ' \
'--paging=[Whether to use a pager when displaying output]:auto|always|never: ' \
'--plus-emph-style=[Style string for emphasized sections of added lines]:STYLE: ' \
'--plus-empty-line-marker-style=[Style string for added empty line marker]:STYLE: ' \
'--plus-non-emph-style=[Style string for non-emphasized sections of added lines that have an emphasized section]:STYLE: ' \
'--plus-style=[Style string for added lines]:STYLE: ' \
'--right-arrow=[Text to display with a changed file path]:STRING: ' \
'--syntax-theme=[The syntax-highlighting theme to use]:SYNTAX_THEME: ' \
'--tabs=[The number of spaces to replace tab characters with]:N: ' \
'--true-color=[Whether to emit 24-bit ("true color") RGB color codes]:auto|always|never: ' \
'--whitespace-error-style=[Style string for whitespace errors]:STYLE: ' \
'-w+[The width of underline/overline decorations]:N: ' \
'--width=[The width of underline/overline decorations]:N: ' \
'--word-diff-regex=[Regular expression defining a '\''word'\'' in within-line diff algorithm]:REGEX: ' \
'--wrap-left-symbol=[End-of-line wrapped content symbol (left-aligned)]:STRING: ' \
'--wrap-max-lines=[How often a line should be wrapped if it does not fit]:N: ' \
'--wrap-right-percent=[Threshold for right-aligning wrapped content]:PERCENT: ' \
'--wrap-right-prefix-symbol=[Pre-wrapped content symbol (right-aligned)]:STRING: ' \
'--wrap-right-symbol=[End-of-line wrapped content symbol (right-aligned)]:STRING: ' \
'--zero-style=[Style string for unchanged lines]:STYLE: ' \
'--24-bit-color=[Deprecated\: use --true-color]:auto|always|never: ' \
'--color-only[Do not alter the input structurally in any way]' \
'--dark[Use default colors appropriate for a dark terminal background]' \
'--diff-highlight[Emulate diff-highlight]' \
'--diff-so-fancy[Emulate diff-so-fancy]' \
'--hyperlinks[Render commit hashes, file names, and line numbers as hyperlinks]' \
'--keep-plus-minus-markers[Prefix added/removed lines with a +/- character, as git does]' \
'--light[Use default colors appropriate for a light terminal background]' \
'-n[Display line numbers next to the diff]' \
'--line-numbers[Display line numbers next to the diff]' \
'--list-languages[List supported languages and associated file extensions]' \
'--list-syntax-themes[List available syntax-highlighting color themes]' \
'--navigate[Activate diff navigation]' \
'--no-gitconfig[Do not read any settings from git config]' \
'--parse-ansi[Display ANSI color escape sequences in human-readable form]' \
'--raw[Do not alter the input in any way]' \
'--relative-paths[Output all file paths relative to the current directory]' \
'--show-colors[Show available named colors]' \
'--show-config[Display the active values for all Delta options]' \
'--show-syntax-themes[Show example diff for available syntax-highlighting themes]' \
'--show-themes[Show example diff for available delta themes]' \
'-s[Display diffs in side-by-side layout]' \
'--side-by-side[Display diffs in side-by-side layout]' \
'-h[Print help (see more with '\''--help'\'')]' \
'--help[Print help (see more with '\''--help'\'')]' \
'-V[Print version]' \
'--version[Print version]' \
'::minus_file -- First file to be compared when delta is being used in diff mode:_files' \
'::plus_file -- Second file to be compared when delta is being used in diff mode:_files' \
&& ret=0
}

(( $+functions[_delta_commands] )) ||
_delta_commands() {
    local commands; commands=()
    _describe -t commands 'delta commands' commands "$@"
}

if [ "$funcstack[1]" = "_delta" ]; then
    _delta "$@"
else
    compdef _delta delta
fi
```

## delta --generate-completion fish
```
complete -c delta -l blame-code-style -d 'Style string for the code section of a git blame line' -r
complete -c delta -l blame-format -d 'Format string for git blame commit metadata' -r
complete -c delta -l blame-palette -d 'Background colors used for git blame lines (space-separated string)' -r
complete -c delta -l blame-separator-format -d 'Separator between the blame format and the code section of a git blame line' -r
complete -c delta -l blame-separator-style -d 'Style string for the blame-separator-format' -r
complete -c delta -l blame-timestamp-format -d 'Format of `git blame` timestamp in raw git output received by delta' -r
complete -c delta -l blame-timestamp-output-format -d 'Format string for git blame timestamp output' -r
complete -c delta -l config -d 'Load the config file at PATH instead of ~/.gitconfig' -r
complete -c delta -l commit-decoration-style -d 'Style string for the commit hash decoration' -r
complete -c delta -l commit-regex -d 'Regular expression used to identify the commit line when parsing git output' -r
complete -c delta -l commit-style -d 'Style string for the commit hash line' -r
complete -c delta -l default-language -d 'Default language used for syntax highlighting' -r
complete -c delta -l diff-stat-align-width -d 'Width allocated for file paths in a diff stat section' -r
complete -c delta -l features -d 'Names of delta features to activate (space-separated)' -r
complete -c delta -l file-added-label -d 'Text to display before an added file path' -r
complete -c delta -l file-copied-label -d 'Text to display before a copied file path' -r
complete -c delta -l file-decoration-style -d 'Style string for the file decoration' -r
complete -c delta -l file-modified-label -d 'Text to display before a modified file path' -r
complete -c delta -l file-removed-label -d 'Text to display before a removed file path' -r
complete -c delta -l file-renamed-label -d 'Text to display before a renamed file path' -r
complete -c delta -l file-style -d 'Style string for the file section' -r
complete -c delta -l file-transformation -d 'Sed-style command transforming file paths for display' -r
complete -c delta -l generate-completion -d 'Print completion file for the given shell' -r -f -a "{bash	'',elvish	'',fish	'',powershell	'',zsh	''}"
complete -c delta -l grep-context-line-style -d 'Style string for non-matching lines of grep output' -r
complete -c delta -l grep-file-style -d 'Style string for file paths in grep output' -r
complete -c delta -l grep-header-decoration-style -d 'Style string for the header decoration in grep output' -r
complete -c delta -l grep-header-file-style -d 'Style string for the file path part of the header in grep output' -r
complete -c delta -l grep-line-number-style -d 'Style string for line numbers in grep output' -r
complete -c delta -l grep-output-type -d 'Grep output format. Possible values: "ripgrep" - file name printed once, followed by matching lines within that file, each preceded by a line number. "classic" - file name:line number, followed by matching line. Default is "ripgrep" if `rg --json` format is detected, otherwise "classic"' -r
complete -c delta -l grep-match-line-style -d 'Style string for matching lines of grep output' -r
complete -c delta -l grep-match-word-style -d 'Style string for the matching substrings within a matching line of grep output' -r
complete -c delta -l grep-separator-symbol -d 'Separator symbol printed after the file path and line number in grep output' -r
complete -c delta -l hunk-header-decoration-style -d 'Style string for the hunk-header decoration' -r
complete -c delta -l hunk-header-file-style -d 'Style string for the file path part of the hunk-header' -r
complete -c delta -l hunk-header-line-number-style -d 'Style string for the line number part of the hunk-header' -r
complete -c delta -l hunk-header-style -d 'Style string for the hunk-header' -r
complete -c delta -l hunk-label -d 'Text to display before a hunk header' -r
complete -c delta -l hyperlinks-commit-link-format -d 'Format string for commit hyperlinks (requires --hyperlinks)' -r
complete -c delta -l hyperlinks-file-link-format -d 'Format string for file hyperlinks (requires --hyperlinks)' -r
complete -c delta -l inline-hint-style -d 'Style string for short inline hint text' -r
complete -c delta -l inspect-raw-lines -d 'Kill-switch for --color-moved support' -r
complete -c delta -l line-buffer-size -d 'Size of internal line buffer' -r
complete -c delta -l line-fill-method -d 'Line-fill method in side-by-side mode' -r
complete -c delta -l line-numbers-left-format -d 'Format string for the left column of line numbers' -r
complete -c delta -l line-numbers-left-style -d 'Style string for the left column of line numbers' -r
complete -c delta -l line-numbers-minus-style -d 'Style string for line numbers in the old (minus) version of the file' -r
complete -c delta -l line-numbers-plus-style -d 'Style string for line numbers in the new (plus) version of the file' -r
complete -c delta -l line-numbers-right-format -d 'Format string for the right column of line numbers' -r
complete -c delta -l line-numbers-right-style -d 'Style string for the right column of line numbers' -r
complete -c delta -l line-numbers-zero-style -d 'Style string for line numbers in unchanged (zero) lines' -r
complete -c delta -l map-styles -d 'Map styles encountered in raw input to desired output styles' -r
complete -c delta -l max-line-distance -d 'Maximum line pair distance parameter in within-line diff algorithm' -r
complete -c delta -l max-line-length -d 'Truncate lines longer than this' -r
complete -c delta -l merge-conflict-begin-symbol -d 'String marking the beginning of a merge conflict region' -r
complete -c delta -l merge-conflict-end-symbol -d 'String marking the end of a merge conflict region' -r
complete -c delta -l merge-conflict-ours-diff-header-decoration-style -d 'Style string for the decoration of the header above the \'ours\' merge conflict diff' -r
complete -c delta -l merge-conflict-ours-diff-header-style -d 'Style string for the header above the \'ours\' branch merge conflict diff' -r
complete -c delta -l merge-conflict-theirs-diff-header-decoration-style -d 'Style string for the decoration of the header above the \'theirs\' merge conflict diff' -r
complete -c delta -l merge-conflict-theirs-diff-header-style -d 'Style string for the header above the \'theirs\' branch merge conflict diff' -r
complete -c delta -l minus-empty-line-marker-style -d 'Style string for removed empty line marker' -r
complete -c delta -l minus-emph-style -d 'Style string for emphasized sections of removed lines' -r
complete -c delta -l minus-non-emph-style -d 'Style string for non-emphasized sections of removed lines that have an emphasized section' -r
complete -c delta -l minus-style -d 'Style string for removed lines' -r
complete -c delta -l navigate-regex -d 'Regular expression defining navigation stop points' -r
complete -c delta -l pager -d 'Which pager to use' -r
complete -c delta -l paging -d 'Whether to use a pager when displaying output' -r
complete -c delta -l plus-emph-style -d 'Style string for emphasized sections of added lines' -r
complete -c delta -l plus-empty-line-marker-style -d 'Style string for added empty line marker' -r
complete -c delta -l plus-non-emph-style -d 'Style string for non-emphasized sections of added lines that have an emphasized section' -r
complete -c delta -l plus-style -d 'Style string for added lines' -r
complete -c delta -l right-arrow -d 'Text to display with a changed file path' -r
complete -c delta -l syntax-theme -d 'The syntax-highlighting theme to use' -r
complete -c delta -l tabs -d 'The number of spaces to replace tab characters with' -r
complete -c delta -l true-color -d 'Whether to emit 24-bit ("true color") RGB color codes' -r
complete -c delta -l whitespace-error-style -d 'Style string for whitespace errors' -r
complete -c delta -s w -l width -d 'The width of underline/overline decorations' -r
complete -c delta -l word-diff-regex -d 'Regular expression defining a \'word\' in within-line diff algorithm' -r
complete -c delta -l wrap-left-symbol -d 'End-of-line wrapped content symbol (left-aligned)' -r
complete -c delta -l wrap-max-lines -d 'How often a line should be wrapped if it does not fit' -r
complete -c delta -l wrap-right-percent -d 'Threshold for right-aligning wrapped content' -r
complete -c delta -l wrap-right-prefix-symbol -d 'Pre-wrapped content symbol (right-aligned)' -r
complete -c delta -l wrap-right-symbol -d 'End-of-line wrapped content symbol (right-aligned)' -r
complete -c delta -l zero-style -d 'Style string for unchanged lines' -r
complete -c delta -l 24-bit-color -d 'Deprecated: use --true-color' -r
complete -c delta -l color-only -d 'Do not alter the input structurally in any way'
complete -c delta -l dark -d 'Use default colors appropriate for a dark terminal background'
complete -c delta -l diff-highlight -d 'Emulate diff-highlight'
complete -c delta -l diff-so-fancy -d 'Emulate diff-so-fancy'
complete -c delta -l hyperlinks -d 'Render commit hashes, file names, and line numbers as hyperlinks'
complete -c delta -l keep-plus-minus-markers -d 'Prefix added/removed lines with a +/- character, as git does'
complete -c delta -l light -d 'Use default colors appropriate for a light terminal background'
complete -c delta -s n -l line-numbers -d 'Display line numbers next to the diff'
complete -c delta -l list-languages -d 'List supported languages and associated file extensions'
complete -c delta -l list-syntax-themes -d 'List available syntax-highlighting color themes'
complete -c delta -l navigate -d 'Activate diff navigation'
complete -c delta -l no-gitconfig -d 'Do not read any settings from git config'
complete -c delta -l parse-ansi -d 'Display ANSI color escape sequences in human-readable form'
complete -c delta -l raw -d 'Do not alter the input in any way'
complete -c delta -l relative-paths -d 'Output all file paths relative to the current directory'
complete -c delta -l show-colors -d 'Show available named colors'
complete -c delta -l show-config -d 'Display the active values for all Delta options'
complete -c delta -l show-syntax-themes -d 'Show example diff for available syntax-highlighting themes'
complete -c delta -l show-themes -d 'Show example diff for available delta themes'
complete -c delta -s s -l side-by-side -d 'Display diffs in side-by-side layout'
complete -c delta -s h -l help -d 'Print help (see more with \'--help\')'
complete -c delta -s V -l version -d 'Print version'
```

dandavison commented:
@plustik thanks a lot for doing this -- would you be able to add some instructions to the manual?

This certainly looks like the right approach to me. I've tried the zsh completions and they seem to work correctly. I'd like to delete the existing completion files rather than distribute them. I guess this may break some downstream packaging scripts, but I still think it's the right thing to do, especially since delta is not at 1.x yet. Anyone, feel free to tell me otherwise.

maverick1872 commented:
@dandavison the other alternative would be to generate the completions as a part of the release artifacts and distribute accordingly. Both would be appropriate in my eyes. It's a question of who has the onus to generate them.

Short term would be a question of retaining backwards compatibility. 

Long term though I agree the completions should be generated at install time.

dandavison commented:
> Short term would be a question of retaining backwards compatibility.
> Long term though I agree the completions should be generated at install time.

Agreed. But I can only think of quite inconvenient ways of communicating the deprecation to downstream maintainers.

OK I've pushed to your branch @plustik: a Makefile target and dumped the output to the existing files.

exploide commented:
As the person who contributed the fish completions in #1165 and @dandavison notified me there, I would like to leave my comment.

I like the approach of generating shell completions automatically when the CLI framework supports it. This helps keeping the completions up to date and results in less manual work in the end. :+1:

I recommend documenting this for package maintainers, so they are aware that they should generate the completion files before packaging delta. However, since you added that to the Makefile, it might be sufficient, I'm not sure.

One possible drawback to consider is that automatically generated completions might be worse than manual completions when it comes to descriptions and argument completion. This can usually be solved by augmenting the CLI definition with additional data when the generation framework supports that. I don't know clap and cannot say if this is the case here. Some examples:

- The description for `--grep-output-type` is ridiculously long. This is not helpful as it gets truncated anyway.
- Argument completion is currently only present for `--generate-completion` itself, but not for other options. In the manual completion script for fish, there are at least completions for:
   - `--inspect-raw-lines` with values: "true false"
   - `--line-fill-method` with values: "ansi spaces"
   - `--paging` with values "auto always never"
   - `--true-color` with values "auto always never"
   - `--syntax-theme` automatically generated with "(delta --list-syntax-themes | cut -f 2)"

I'm not saying this is a must have for the first attempt of generated completions, but I think it would be great if these could be added eventually. I know that some Go tools which use the Cobra framework, like podman or docker have awesome argument completion. I could imagine that clap is also capable of this, but I don't know how much work it is.

plustik commented:
> One possible drawback to consider is that automatically generated completions might be worse than manual completions when it comes to descriptions and argument completion. This can usually be solved by augmenting the CLI definition with additional data when the generation framework supports that. I don't know clap and cannot say if this is the case here. Some examples:
> 
>     * The description for `--grep-output-type` is ridiculously long. This is not helpful as it gets truncated anyway.
> 
>     * Argument completion is currently only present for `--generate-completion` itself, but not for other options. In the manual completion script for fish, there are at least completions for:
>       
>       * `--inspect-raw-lines` with values: "true false"
>       * `--line-fill-method` with values: "ansi spaces"
>       * `--paging` with values "auto always never"
>       * `--true-color` with values "auto always never"
>       * `--syntax-theme` automatically generated with "(delta --list-syntax-themes | cut -f 2)"
> 
> 
> I'm not saying this is a must have for the first attempt of generated completions, but I think it would be great if these could be added eventually. I know that some Go tools which use the Cobra framework, like podman or docker have awesome argument completion. I could imagine that clap is also capable of this, but I don't know how much work it is.

I was able to fix most of your examples and while the long description for `--grep-output-type` isn't perfect it's not a big drawback from my point of view.
But I don't think there is a good solution for completing `--syntax-theme` without editing the completion files after generation.

My point is: The automatic generation can definitely be improved further by adding more information to the CLI definition, but don't think clap allows us to generate completion files as good as hand-written files can be.
In the end it's a decision between maintaining perfect completion files by updating them manually after every change to the CLI and generating good enough files automatically.

@dandavision I added an entry to the manual. I'm not sure, whether I included all relevant information. If you'd expect anything else in that page, I can add that.
I think there are cleaner solutions for the argument completions I fixed in [#e516736](https://github.com/dandavison/delta/pull/1561/commits/e5167365b5c04e322b5ea2a6ba48180ce7fcbb2a), but they would require changing the types of the corresponding attributes in `cli::Opt`, which would mess with the `set_options` macro. I don't know much about the codebase, but I don't think it's worth the effort.

dandavison commented on src/cli.rs:
```diff
@@ -595,7 +601,7 @@ pub struct Opt {
     pub line_buffer_size: usize,
 
-    #[arg(long = "line-fill-method", value_name = "STRING")]
+    #[arg(long = "line-fill-method", value_name = "STRING", value_parser = ["ansi spaces"])]
```
There's a test failure I think because this should be split into two words.

dandavison replied:
(I pushed the fix)

dandavison commented:
Thanks very much @plustik, and @exploide and @maverick1872. I've merged this, with the auto-generated completions in the repo.
